client/remoterepo.c

```diff
@@ -106,11 +113,490 @@ _handle_curl_cb(
+       !ml_file)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    fprintf(stdout, "Validating metalink (%s)...\n", pszFile);
+    if(ml_file->digest == NULL)
+    {
+        fprintf(stderr,
+                "Validating metalink (%s) FAILED (digest missing)\n", pszFile);
+        dwError = ERROR_TDNF_CHECKSUM_VALIDATION_FAILED;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+    dwError = TDNFCheckHash(pszFile, ml_file->digest, ml_file->type);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+cleanup:
+    if(ml_file->digest)
+    {
+        TDNF_SAFE_FREE_MEMORY(ml_file->digest);
+    }
+    if(ml_file->type)
+    {
+        TDNF_SAFE_FREE_MEMORY(ml_file->type);
+    }
+    if(ml_file->filename)
+    {
+        TDNF_SAFE_FREE_MEMORY(ml_file->filename);
+    }
+    if(ml_file)
+    {
+        TDNF_SAFE_FREE_MEMORY(ml_file);
+    }
+    return dwError;
+error:
+    goto cleanup;
+}
+
+uint32_t
+TDNFCheckHash(
+    const char *filename,
+    const char *digest,
+    const char *type
+    )
+{
+    uint32_t dwError = 0;
+    unsigned char *result;
+    metalink_digest_alias *digest_alias;
+
+    if(IsNullOrEmptyString(filename) ||
+       IsNullOrEmptyString(digest)   ||
+       IsNullOrEmptyString(type))
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+    for(digest_alias = digest_aliases; digest_alias->alias_name;
+        ++digest_alias)
+    {
+        if(!strcmp(digest_alias->alias_name, type))
+        {
+            digest_alias->func(filename, &result);
+            if(!memcmp(result, digest,
+                       digest_alias->length))
+            {
+                /* sha*sum style verdict output */
+                fprintf(stdout, "Validating metalink (%s) OK\n", filename);
+            }
+            else
+            {
+                fprintf(stderr, "Validating metalink (%s) FAILED (digest mismatch)\n",
+                        filename);
+                dwError = ERROR_TDNF_INVALID_REPO_FILE;
+                BAIL_ON_TDNF_SYSTEM_ERROR(dwError); 
+            }
+            break;
+        }
+    }
+
+cleanup:
+    if(result)
+    {
+        TDNF_SAFE_FREE_MEMORY(result);
+    }
+    return dwError;
+error:
+    goto cleanup;
+}
+
+/* Returns nonzero if hex_digest is properly formatted; that is each
+   letter is in [0-9A-Za-z] and the length of the string equals to the
+   result length of digest * 2. */
+uint32_t
+TDNF_Check_Hex_Digest
+    (const char *hex_digest,
+    int digest_length
+    )
+{
+    int i = 0;
+    if(!hex_digest ||
+       !digest_length)
+    {
+        return 0;
+    }
+    for(i = 0; hex_digest[i]; ++i)
+    {
+        char c = hex_digest[i];
+        if(!(('0' <= c && c <= '9') || ('a' <= c && c <= 'z') ||
+             ('A' <= c && c <= 'Z')))
+        {
+            return 0;
+        }
+    }
+    return digest_length * 2 == i;
+}
+
+char 
+TDNF_HexToUint
+    (const char *s)
+{
+    char buf[3];
```

Comment:
init to 0, check for null of function param and style mismatch in function braces

Reply:
Sure.